### ISSUE TITLE
Linux middle mouse paste inserts text twice

### ISSUE THREAD

**Issue by cloos** [atom/atom#8648]
- mark somewhere text with the mouse
- File -> New File
- press middle mouse button to insert text

The text is inserted twice.

This happens since 1.0.8 and even in the latest 1.0.10.

atom-amd64.deb
Ubuntu 15.04

**Comment by peduxe** [comment]
[Here is a solution.](https://atom.io/packages/sublime-style-column-selection)

**Comment by cloos** [comment]
If I install 1.0.7, the text is inserted only once, so everything is fine.
The issue appeared since 1.0.8.

As this seams to be a core issue, I don't think installing a community package is a right solution.

Also I think that this wouldn't solve the issue, as by "mark somewhere text with the mouse" I mean mark text for example inside the browser, and not text inside the atom editor.

**Comment by peduxe** [comment]
I couldn't reproduce this. It just pasted once.

**Comment by nicksanders** [comment]
I'm experiencing this too after upgrading from 1.0.7 to 1.0.11

**Comment by dpanter** [comment]
Could not reproduce this issue with 1.0.11, fresh install.
Following OP instructions, it only pasted once.
Linux Mint Debian Edition, KDE4.

**Comment by nicksanders** [comment]
This is still happening with Atom 1.0.13

To reproduce
mv ~/.atom ~/.atom-saved
atom --safe

In untitled tab 
type "test" then press enter twice
double click left mouse button on test to highlight
move mouse to line 3 and press middle mouse button

![atom1](https://cloud.githubusercontent.com/assets/809448/9961499/c512ff7a-5e18-11e5-9227-d6db3513476d.png)

**Comment by peduxe** [comment]
@nicksanders I can confirm this happens to me as well.

i3wm, Linux 15.04 64-bits Atom 1.0.14

**Comment by dpanter** [comment]
@nicksanders I'll revise my earlier statement and say I can in fact reproduce this exact behavior as well, a left click before pasting with a middle click changes the result.

However... as I tried middle clicking more than once, I realized each subsequent middle click makes my 1.0.11 doubly paste what it just pasted. Preceding left click or not makes no difference.
I get "test" -> "testtest" -> "testtesttesttest" -> "testtesttesttesttesttesttesttest" -> "testtesttesttesttesttesttesttesttesttesttesttesttesttesttesttest" and so on.

Probably not the intended behavior.

**Comment by hainesr** [comment]
This is happening in version 1.0.19 (from the PPA) on Ubuntu 14.04 too.

**Comment by slegga** [comment]
I also have this problem on Atom 1.0.19 kubuntu 15.04 32 bit.
If before pressing middle button  I do a double right click, the middle button works with no extra paste.

**Comment by nicksanders** [comment]
This is still happening with Atom 1.1.0

**Comment by JohnA3** [comment]
Still here with Atom 1.3.0b1 on linux64

**Comment by pedantic-git** [comment]
Agreed. Happening to me too with 1.2.0 official build (Ubuntu 15.10 64-bit).

**Comment by jetole** [comment]
Is this issue still open or has it been resolved? I am experiencing this bug as well.

**Comment by nicksanders** [comment]
Still happening to me as well

**Comment by 50Wliu** [comment]
> Is this issue still open or has it been resolved?

It's still open.

**Comment by oetiker** [comment]
same here

**Comment by hainesr** [comment]
Interestingly, Atom appears to be seeing this double paste as two separate operations; a single undo (ctrl-z) only removes the second one.

This has become my workaround but it would be nice to have this fixed.

**Comment by iivvoo** [comment]
Same problem here, very consistent (and very annoying). Atom 1.3.2 on Ubuntu 14.04 x64

But it doesn't always happen. Pastes on empty lines seem to go well but inside an existing line / piece of text usually results in double pastes.

**Comment by jlsalvador** [comment]
I made an horrific-patch that prevents duplicates (I have not time to investigate the source):

```
diff --git a/src/text-editor-component.coffee b/src/text-editor-component.coffee
index 430b0c0..4e08c49 100644
--- a/src/text-editor-component.coffee
+++ b/src/text-editor-component.coffee
@@ -34,6 +34,7 @@ class TextEditorComponent
   gutterComponent: null
   mounted: true
   initialized: false
+  lastPasteSelectionClipboardEvent: null
 
   Object.defineProperty @prototype, "domNode",
     get: -> @domNodeValue
@@ -638,8 +639,9 @@ class TextEditorComponent
       Math.pow(scrollDelta / 2, 3) / 280
 
     pasteSelectionClipboard = (event) =>
-      if event?.which is 2 and process.platform is 'linux'
+      if event?.which is 2 and process.platform is 'linux' and @lastPasteSelectionClipboardEvent != event
         if selection = require('./safe-clipboard').readText('selection')
+          @lastPasteSelectionClipboardEvent = event
           @editor.insertText(selection)
 
     window.addEventListener('mousemove', onMouseMove)
```

**Comment by pedantic-git** [comment]
Related to this: I was looking at a package page under "Settings -> Updates" and I noticed that left-clicking "CHANGELOG" opened the changelog in two tabs, and clicking on the package address opened two tabs in my browser with the package page. I suspect it's connected.

**Comment by rednil** [comment]
I am experiencing this as well in about 50% of the cases. However, I believe this is an ubuntu issue, since I see this in gvim as well, but NOT in the terminal, for example. It happened in 14.04, and it still happens in 15.10

**Comment by jperelli** [comment]
This happens to me too. It's also about 75% of the cases.

**Comment by person666** [comment]
@rednil it is not an Ubuntu issue I get the issue on arch gnome and kde

**Comment by beastea** [comment]
Experiencing issue with pasting on middle button click on Fedora23 x86_64. It works for me fine with atom 1.4.0 version but completely breaks starting from 1.4.1

**Comment by dpanter** [comment]
Upgraded to 1.4.3 today. This is still broken.
However, if I move the mouse a bit between pasting, it only pastes once.

**Comment by belluzj** [comment]
> However, if I move the mouse a bit between pasting, it only pastes once.

Same for me: if I left-click somewhere, then without moving the mouse, I middle-click to paste, I get the text pasted twice. Conversely, if I directly middle-click somewhere without left-clicking first, I get only one pasting.

**Comment by jetole** [comment]
I'm surprised an issue of this magnitude has been pending for so long. Does
anyone know if this is being addressed? Any ETA on when it may be fixed?

On Mon, Feb 15, 2016 at 5:46 AM, belluzj <notifications@github.com> wrote:

> However, if I move the mouse a bit between pasting, it only pastes once.
>
> Same for me: if I left-click somewhere, then without moving the mouse, I
> middle-click to paste, I get the text pasted twice. Conversely, if I
> directly middle-click somewhere without left-clicking first, I get only one
> pasting.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/issues/8648#issuecomment-184162740>.
>



-- 

:: Sincerely,
:: Joseph Tole

**Comment by person666** [comment]
@jetole I agree this has affected my daily use of atom since I switched to it. It is kinda strange it has not been fixed yet.

**Comment by dennisjac** [comment]
I'm seeing this two using the official RPMs on a Fedora 22 system.

**Comment by TeChn4K** [comment]
Same here ...
(Need +1 github feature ...)

**Comment by jetole** [comment]
This bug is still present on 1.6.0.

On Thu, Mar 10, 2016 at 10:16 AM, TeChn4K <notifications@github.com> wrote:

> Same here ...
> (Need +1 github feature ...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/issues/8648#issuecomment-194898506>.
>



-- 

:: Sincerely,
:: Joseph Tole

**Comment by TeChn4K** [comment]
**Workaround :** 

Before middle click, do not click left (to select your line). Otherwise moove your mouse before.

ty @dpanter and @belluzj

That is a very annoying bug :/

**Comment by iivvoo** [comment]
I've taken a stab at this because it really keeps annoying me.

My guess is that the problem is caused by the fact that if "dragging stops", paseSelectionClipboard is called. And the middle-mouse-pasting might be handled as a mini-drag (if you move just a tiny bit)

So not pasting if we're stopping dragging solves the problem for me. I'm just not sure what if might break. I've created a PR, let's see what happens with it :)

PR: https://github.com/atom/atom/pull/11262

**Comment by person666** [comment]
@iivvoo  Does your fix work? This issue has been around for ages and is consistently annoyingly showing up in daily work.

**Comment by iivvoo** [comment]
@person666 It worked for me though I'm currently using stable releases again so the issue has returned. I've learned to somewhat work around the issue - using vim-mode copy/paste more of wiggle my mouse a bit before pasting. It would be nice though if someone could review the PR

**Comment by polefishu** [comment]
This bug is still present on 1.12.2 on ubuntu 16.04 gnome, and  first click middle mouse the paste inserts text twice, immediately click again  the paste inserts text three times...

**Comment by TristanLefebure** [comment]
Same issue, very annoying, makes Atom nearly unusable in my opinion

(Ubuntu 16.04, Atom 1.7.3)

**Comment by iivvoo** [comment]
Quirks like this in Atom and the fact that the bug and my PR are completely ignored made me move to vscode. Problem solved for me :)

**Comment by psyer** [comment]
I just started using Atom (1.12.7) as an editor, and this bug is driving me nuts. This simple daily operation makes this editor unusable. @iivvoo  has put in a PR (thank you!) to fix this. For the love of the universe please merge it and test it. On to look at other editors I guess...

**Comment by willigott** [comment]
Still experience this issue and would be very glad if it could be fixed.

**Comment by nicksanders** [comment]
Given up waiting, changed to VS Code :)

**Comment by mkwinco** [comment]
Same here, ubuntu 16.04. I really do want to try Atom, but this makes me think twice about it...

**Comment by wdecoster** [comment]
Annoying as hell, I would really appreciate it if this issue could be solved... (Ubuntu 16.04)

**Comment by dietmar** [comment]
Everyone, no hard feelings, but please stop adding comments like "I also have this issue" and "it annoys me", they are not helpful. This issue has been confirmed, it has received the label "bug" and also the label "help-wanted", so the situation is pretty clear: everyone is invited to contribute towards resolving the issue.

If people's inboxes get hammered with such meaningless comments, they will unsubscribe from this issue and it will be left entirely in the dark. So let's please stick to discussing the issue in the direction of actually resolving it.

**Comment by sleeksorrow** [comment]
Sadly the person adding the "help-wanted" tag forgot to tell which kind of help is wanted. There is a clear reproduction scenario in the comments. Short version: 1) Mouse selection copy, 2) click in atom window, 3) mmb paste. Really it's that easy.

So the most obvious help people can provide is to tell that they are affected, so developers know that newer distribution versions did not fix the issue magically.

So if the developers need more help/info/money/coffee, we need clear directions. I'm happy to assist.

**Comment by dietmar** [comment]
@sleeksorrow the way forward is pretty obvious: all of the source code is openly available, and the problem has been described with sufficient detail; someone needs to dive into the problem, actually solve it, test the solution, and provide the code changes that fix the problem. In short, a pull request.

If you don't have time to do that, that's fine, but please understand that a thousand people saying "I want this fixed but I don't want to do it myself" doesn't move anything forward.

**Comment by sleeksorrow** [comment]
The pull request from @iivvoo is linked up there and is being ignored since one year now. I don't see any reason why I should make a duplicate pull request.

**Comment by jlsalvador** [comment]
@dietmar 
- https://github.com/atom/atom/issues/8648#issuecomment-172554041
- https://github.com/atom/atom/pull/11262

**Comment by 50Wliu** [comment]
That PR is distinctly labeled as a "workaround" and not an actual solution.

**Comment by schlomo** [comment]
+1

**Comment by sleeksorrow** [comment]
@dietmar I really appreciate your work on this issue. Thank you very much.

**Comment by 50Wliu** [comment]
Thanks @dietmar for your initiative!  The fix is scheduled to be released in Atom 1.17, soon to be in beta.

**Comment by Asraelite** [comment]
This issue has appeared for me again in Atom 1.19.0.beta1-1, although it is not in 1.18.

**Comment by hex007** [comment]
I can affirm that that Atom 1.19.0.beta2 also has this bug

**Comment by ungb** [comment]
@

**Comment by schlomo** [comment]
Sorry, I switched to Visual Studio Code because of this bug and because it loads *much* faster and because it doesn't clobber Makefiles :-(

**Comment by xenadu000** [comment]
This issue persists in Ubuntu 17.10. I found this thread while looking for a solution.

**Comment by wdecoster** [comment]
Interestingly enough, this issue doesn't affect me anymore.

Ubuntu 17.10
Atom    : 1.22.1
Electron: 1.6.15
Chrome  : 56.0.2924.87
Node    : 7.4.0
